rpc, test: add `sendmsgtopeer` rpc and a test for net-level deadlock situation

mzumsande opened the pull request:
This adds a `sendmsgtopeer` rpc (for testing only) that allows a node to send a message (provided in hex) to a peer.
While we would usually use a `p2p` object instead of a node for this in the test framework, that isn't possible in situations where this message needs to trigger an actual interaction of multiple nodes.

Use this rpc to add test coverage for the bug fixed in #27981 (that just got merged):
The test lets two nodes (almost) simultaneously send a single large (4MB) p2p message to each other, which would have caused a deadlock previously (making this test fail), but succeeds now.

As can be seen from the discussion in #27981, it was not easy to reproduce this bug without `sendmsgtopeer`. I would imagine that `sendmsgtopeer` could also be helpful in various other test constellations.

mzumsande committed:
test: add functional test for deadlock situation

mzumsande committed:
rpc: add test-only sendmsgtopeer rpc

This rpc can be used when we want a node to send a message, but
cannot use a python P2P object, for example for testing of low-level
net transport behavior.

brunoerg commented:
Concept ACK

ajtowns commented:
Concept ACK. Probably should add some basic functional tests to rpc_net.py with a `P2PInterface` checking that it works -- what happens if `msg_type` is 0 bytes, 13 bytes (`> protocol.h:COMMAND_SIZE`); what if `msg` is empty or more than 4MB (`net.h:MAX_PROTOCOL_MESSAGE_LENGTH`) or 32MB (`serialize.h:MAX_SIZE`)?

maflcko reviewed (approved):
lgtm. Feel free to ignore the nits.

brunoerg commented on src/rpc/net.cpp:
```diff
@@ -968,6 +968,53 @@ static RPCHelpMan addpeeraddress()
+{
+    const NodeId peer_id{request.params[0].getInt<int64_t>()};
+    const std::string& msg_type{request.params[1].get_str()};
+    if( msg_type.size() > CMessageHeader::COMMAND_SIZE) {
```
nit
```suggestion
    if (msg_type.size() > CMessageHeader::COMMAND_SIZE) {
```

mzumsande replied:
fixed

russeree commented:
Concept ACK this would be useful.

brunoerg commented on test/functional/p2p_net_deadlock.py:
```diff
@@ -0,0 +1,46 @@
+    def run_test(self):
+        node0 = self.nodes[0]
+        node1 = self.nodes[1]
+        self.connect_nodes(0, 1)
```
I think you don't need to override `setup_network`:
```diff
diff --git a/test/functional/p2p_net_deadlock.py b/test/functional/p
2p_net_deadlock.py
index 1b2139cc8..4ed9170e4 100755
--- a/test/functional/p2p_net_deadlock.py
+++ b/test/functional/p2p_net_deadlock.py
@@ -17,14 +17,9 @@ class NetDeadlockTest(BitcoinTestFramework):
         self.setup_clean_chain = True
         self.num_nodes = 2
 
-    def setup_network(self):
-        self.setup_nodes()
-
     def run_test(self):
         node0 = self.nodes[0]
         node1 = self.nodes[1]
-        self.connect_nodes(0, 1)
-        self.sync_all()
 
         self.log.info("Simultaneously send a large message on both sides")
         rand_msg = random_bytes(4000000).hex()
```

mzumsande replied:
done

mzumsande commented on src/rpc/net.cpp:
```diff
@@ -968,6 +968,53 @@ static RPCHelpMan addpeeraddress()
+    return ret;
+},
+    };
+}
```
For some reason I always thought that clang-format doesn't apply to rpc indentation (because most of the existing ones don't comply with it), but that doesn't seem to be the case, so I've applied it.

mzumsande commented on src/rpc/net.cpp:
```diff
@@ -968,6 +968,53 @@ static RPCHelpMan addpeeraddress()
+    return RPCHelpMan{"sendmsgtopeer",
+                "\nSend a p2p message to a peer specified by id.\n"
+                "The message type and body must be provided, the message header will be generated"
+                "This RPC is for testing only\n",
```
done

mzumsande commented on src/rpc/net.cpp:
```diff
@@ -968,6 +968,53 @@ static RPCHelpMan addpeeraddress()
+    if( msg_type.size() > CMessageHeader::COMMAND_SIZE) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Error: msg_type too long, max is (%i)", CMessageHeader::COMMAND_SIZE));
+    }
+    const std::vector<unsigned char> msg{ParseHex(request.params[2].get_str())};
```
done, and added a `JSONRPCError` if parsing fails.

mzumsande commented on test/functional/p2p_net_deadlock.py:
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env python3
+# Copyright (c) 2023-The Bitcoin Core developers
```
used `2023-present` as suggested

mzumsande commented on test/functional/p2p_net_deadlock.py:
```diff
@@ -0,0 +1,46 @@
+        rand_msg = random_bytes(4000000).hex()
+
+        thread0 = threading.Thread(target=send_message, args=(node0, 0, "unknown", rand_msg))
+        thread1 = threading.Thread(target=send_message, args=(node1, 0, "unknown", rand_msg))
```
thanks, used `target=node1.sendmsgtopeer` and removed the wrapper

mzumsande commented:
Thanks for the reviews! I pushed an update with changes due to feedback, haven't @ajtowns suggestion yet:

> Concept ACK. Probably should add some basic functional tests to rpc_net.py with a `P2PInterface` checking that it works -- what happens if `msg_type` is 0 bytes, 13 bytes (`> protocol.h:COMMAND_SIZE`); what if `msg` is empty or more than 4MB (`net.h:MAX_PROTOCOL_MESSAGE_LENGTH`) or 32MB (`serialize.h:MAX_SIZE`)?

Will add this to `rpc_net.py` soon. I think the best approach might be to allow these kinds of things things (e.g. empty `msg_type`, message body > 4MB) even if they don't make much sense - as long as they can be dealt with locally and don't trigger any asserts, like the oversized `msg_type` does.

> Error: RPC command "sendmsgtopeer" not found in RPC_COMMANDS_SAFE_FOR_FUZZING or RPC_COMMANDS_NOT_SAFE_FOR_FUZZING. Please update test/fuzz/rpc.cpp.

updated with the same comment `getblockfrompeer` has, indicating that no msg will be sent when there are no peers.

brunoerg reviewed (approved):
utACK d1404b344d5907d159502a2cb05aa0ad557ebb8f

mzumsande committed:
test: add basic tests for sendmsgtopeer to rpc_net.py

mzumsande commented:
I added a commit for basic coverage in `rpc_net.py` and made some other minor changes.

> what if `msg` is empty or more than 4MB (`net.h:MAX_PROTOCOL_MESSAGE_LENGTH`) or 32MB (`serialize.h:MAX_SIZE`)?

I added coverage for >4MB. I couldn't test the case >32MB, because already at ~16MB I get a failure passing a string that long via RPC (independent on the actual RPC, so this can't be handled within `sendmsgtopeer`).

ajtowns commented:
ACK b3a93b409e7fb33af77bd34a269a3eae71d3ba83

LGTM. `bitcoin-cli sendmsgtopeer 0 mempool ""` to a peer that has you whitelisted works as a poor man's `importmempool` (#27460) which is neat. Confirm that `p2p_net_deadlock` test catches the bug if #27981 is reverted (and running `netstat -n` confirms the tcp buffers are just sitting there being ignored while the test case is running).

achow101 commented:
ACK b3a93b409e7fb33af77bd34a269a3eae71d3ba83
